Update storage docs for ink! v4 (#137)
paritytech/ink-docs: versioned_docs/version-4.0.0-alpha.1/datastructures/storage-in-metadata.md

slug: /datastructures/storage-in-metadata
---

The storage layout of a contract is reflected inside the metadata. It allows third-party 
tooling to work with contract storage and can also help to better understand the storage 
layout of any given contract.

Given a contract with the following storage:

```rust
#[ink(storage)]
pub struct MyContract {
    balance: Balance,
    block: BlockNumber,
    something: Lazy<bool>,
}
```

The storage will be reflected inside the metadata as like follows:

```json
"root": {
  "layout": {
    "struct": {
      "fields": [
        {
          "layout": {
            "leaf": {
              "key": "0x00000000",
              "ty": 0
            }
          },
          "name": "balance"
        },
        {
          "layout": {
            "leaf": {
              "key": "0x00000000",
              "ty": 1
            }
          },
          "name": "block"
        },
        {
          "layout": {
            "root": {
              "layout": {
                "leaf": {
                  "key": "0xb1f4904e",
                  "ty": 2

>                 }

Reviewer: Not sure in how much detail we want to talk about this in the docs, but there are more keys than just `root` and `leaf` available to describe items in storage. E.g we can have a root key that contains an `array` or `enum` layout

Author: Hmm yeah, I think it is a good point but I failed to bring up an example covering everything that is also small and easy to digest...

              },
              "root_key": "0xb1f4904e"
            }
          },
          "name": "something"
        }
      ],
      "name": "MyContract"
    }
  },
  "root_key": "0x00000000"
}
```

> We observe that the storage is layed out as a tree, where tangible storage values end up 

Reviewer 2: ```suggestion
We observe that the storage layout is represented as a tree, where tangible storage values end up 
```

inside a `leaf`. Because of `Packed` encoding, leafs can share the same storage key, and 
in order to reach them you'd need fetch and decode the whole storage cell under this key.

> 

Author: Good idea! I think our illustrator is already working on some kind of storage layout visualization

A `root_key` is meant to either be used to directly access a `Packed` storage field 
or to serve as the base key for calculting the actual keys needed to access values in 

> `Non-Packed` fields (such as Mappings).

Reviewer 2: ```suggestion
non-`Packed` fields (such as `Mapping`s).
```

Base storage keys are always 4 bytes in size. However, the storage API of the contracts 
pallet supports keys of arbitrary length. In order to reach a mapping value, the storage 
key of said value is calculated at runtime.

The formula to calculate the base storage key `S` used to access a mapping value under the 
key `K` for a mapping with base key `B` can be expressed as follows:

```
S = B + scale::encode(K)
```
 
In words, find the base (root) key of the mapping and concatenate it with the 
SCALE encoded key of the mapped value to obtain the actual storage key used to 
access the mapped value.

## Accessing storage items with the `contractsApi` runtime call API

There are two ways to query for storage fields of smart contracts from outside a contract. 
Both methods are accessible via the [`polkadot-js`](https://polkadot.js.org/apps/) web UI.

The straight forward way to query a contracts storage is via a 
[`runtime API`](https://polkadot.js.org/apps/#/runtime) call, using the `contractsApi` 
endpoint provided by the contracts pallet. The endpoint provides a `getStorage` method, 
which just expects a contract address and a storage key as arguments.

> For example, to access the root storage struct under the key `0x00000000` of a contrect, 

Reviewer 2: ```suggestion
For example, to access the root storage struct under the key `0x00000000` of a contract, 
```

> just specify the contracts address and the storage key `0x00000000` as-is. The API call 

Reviewer 2: ```suggestion
just specify the contract's address and the storage key `0x00000000` as-is. The API call 
```

will return the scale-encoded root storage struct of the contract.

## Accessing storage items with the `childState` RPC call API

> Under the hood, each contract gets it's own

Reviewer 2: ```suggestion
Under the hood, each contract gets its own
```

- The child trie ID of the contract, represented as a `PrefixedStorageKey`
- The hashed storage key of the storage field

### Finding the contracts child trie ID

The child trie ID is the `Blake2_256` hash of the contracts instantiation nonce
concatenated to it's `AccountId`. You can find it in [`polkadot-js chainstate query interface`](https://polkadot.js.org/apps/#/chainstate), where you need to execute the
`contracts_contractInfoOf` state query.

It can also be calculate manually according to the following code snippet. The 
instantiation note of the contract must be still be know. You can get it using the 
`contracts_nonce` chain state query in polkadot-js UI.

```rust
use sp_core::crypto::Ss58Codec;
use parity_scale_codec::Encode;

// Given our contract ID is 5DjcHxSfjAgCTSF9mp6wQBJWBgj9h8uh57c7TNx1mL5hdQp4
let account: AccountId32 =
    Ss58Codec::from_string("5DjcHxSfjAgCTSF9mp6wQBJWBgj9h8uh57c7TNx1mL5hdQp4").unwrap();
// Given our instnatiation nonce was 1
let nonce: u64 = 1;

// The child trie ID can be calculated as follows:
let trie_id = (&account, nonce).using_encoded(Blake2_256::hash);
```

### Calcualte the `PrefixedStorageKey` from the child trie ID

> A [`PrefixedStorageKey`](https://docs.rs/sp-storage/10.0.0/sp_storage/struct.PrefixedStorageKey.html) 

Author: Good catch

              "key": "0x00000000",
              "ty": 1
            }
          },
          "name": "block"
        },
        {
          "layout": {
            "root": {
              "layout": {
                "leaf": {
                  "key": "0xb1f4904e",
                  "ty": 2
                }
              },
              "root_key": "0xb1f4904e"
            }
          },
          "name": "lazy"
        }
      ],
      "name": "MyContract"
    }
  },
  "root_key": "0x00000000"
}
```

We observe that the storage layout is represented as a tree, where tangible storage values 
end up inside a `leaf`. Because of 
[`Packed`](https://docs.rs/ink_storage_traits/4.0.0-beta.1/ink_storage_traits/trait.Packed.html) 
encoding, leafs can share the same storage key, and 
in order to reach them you'd need fetch and decode the whole storage cell under this key.

A `root_key` is meant to either be used to directly access a `Packed` storage field 
or to serve as the base key for calculating the actual keys needed to access values in 
non-`Packed` fields (such as `Mapping`s).
Layouts under a root key can contain `leaf`s directly, but also `struct`, `enum` or `array` 

> layout, which are used 

Reviewer 2: The sentence cut off here is strange.

Author: This wasn't supposed to go in :see_no_evil:

slug: /datastructures/storage-in-metadata
---

The storage layout of a contract is reflected inside the metadata. It allows third-party 

> tooling to work with contract stroage and can also help to better understand the storage 

Reviewer 3: ```suggestion
tooling to work with contract storage and can also help to better understand the storage 
```

        }
      ],
      "name": "MyContract"
    }
  },
  "root_key": "0x00000000"
}
```

We observe that the storage is layed out as a tree, where tangible storage values end up 
inside a `leaf`. Because of `Packed` encoding, leafs can share the same storage key, and 
in order to reach them you'd need fetch and decode the whole storage cell under this key.

A `root_key` is meant to either be used to directly access a `Packed` storage field 
or to serve as the base key for calculting the actual keys needed to access values in 
`Non-Packed` fields (such as Mappings).

## Storage key calculation for mappings

Root storage keys are always 4 bytes in size. However, the storage API of the contracts 
pallet supports keys of arbitrary length. In order to reach a mapping value, the storage 
key of said value is calculated at runtime.

The formula to calculate a storage key `S` used to access a mapping value under the key `K`
for a mapping with base key `B` can be expressed as follows:

```
S = B + scale::encode(K)
```
 
In words, find the base (root) key of the mapping and concatenate it with the 
SCALE encoded key of the mapped value to obtain the actual storage key used to 

> access the mapped value.

Reviewer 2: Isn't there a hashing done as well?

Author: Oh, good catch! Yes, IIRC in the latest versions of pallet contracts the key is blake2 hashed and append to its own hash to that the key again. You are very right, if you were try to access storage off-chain (without going through the pallet storage API) , I should absolutely documents that here!

Reviewer 3: <img width="650" alt="image" src="https://user-images.githubusercontent.com/18346821/215488461-c83c0ad7-7338-481e-b024-6f7d1f3bc98a.png">

I want to highlight that a new version uses `Blake2_128Concat`. I don't know whether we need to mention it in the documentation or not.

Author: I was also asking myself if it does really fit into the docs. But I came to the conclusion that it should be documented, as the storage layout is part of the contract metadata. I already just wrote a few paragraphs about it. Figuring out how this exactly works requires reading through a lot of substrate code otherwise.
